File: Linked_List/intersecting_node/intersecting_node.ts
/**
 * This is complete function of Leetcode in TS
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function getIntersectionNode(headA: ListNode | null, headB: ListNode | null): ListNode | null {
    let st = new Set<ListNode>();
    while(headA) {
        st.add(headA);
        headA = headA.next;
    }
    while(headB) {
        if(st.has(headB)) return headB;
        st.add(headB);
        headB = headB.next;
    }
    return null;

};

File: Recursion/intersecting_node.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        int lenA = length(headA);
        int lenB = length(headB);

        while (lenA > lenB) {
            headA = headA.next;
            lenA--;
        }
        while (lenB > lenA) {
            headB = headB.next;
            lenB--;
        }

        return findIntersection(headA, headB);
    }

    int length(ListNode head) {
        if (head == null) return 0;
        return 1 + length(head.next);
    }

    ListNode findIntersection(ListNode a, ListNode b) {
        if (a == null || b == null) return null;
        if (a == b) return a;
        return findIntersection(a.next, b.next);
    }
}
